src/mpid/common/genq/mpidu_genqi_shmem_types.h

```diff
@@ -9,13 +9,43 @@
 #include "mpl.h"
 #include "mpidu_init_shm.h"
 
+typedef void *MPIDU_genq_shmem_pool_t;
+typedef void *MPIDU_genq_shmem_queue_t;
+
 typedef struct MPIDU_genqi_shmem_cell_header {
     uintptr_t handle;
-    MPL_atomic_int_t in_use;
-    uintptr_t next;
-    uintptr_t prev;
+    int block_idx;
+    union {
+        struct {
+            uintptr_t next;
+        } serial_queue;
+        struct {
+            uintptr_t prev;     /* used in the tail queue */
+            uintptr_t next;     /* used in the detached header queue */
+        } inverse_queue;
+        struct {
+            MPL_atomic_ptr_t next_m;
+        } nem_queue;
+    } u;
 } MPIDU_genqi_shmem_cell_header_s;
 
+typedef union MPIDU_genq_shmem_queue {
+    struct {
+        union {
+            uintptr_t s;
+            MPL_atomic_ptr_t m;
+            uint8_t pad[MPIDU_SHM_CACHE_LINE_LEN];
+        } head;
+        union {
+            uintptr_t s;
+            MPL_atomic_ptr_t m;
+            uint8_t pad[MPIDU_SHM_CACHE_LINE_LEN];
+        } tail;
+        unsigned flags;
+    } q;
+    uint8_t pad[3 * MPIDU_SHM_CACHE_LINE_LEN];
+} MPIDU_genq_shmem_queue_u;
+
```

Comment:
The queue union should be exposed in the public `mpidu_genq_shmem_queue.h`. Otherwise, the user would have not idea how much memory is needed for the queue object.

Reply:
User doesn't need know, the compiler knows :) , and that is sufficient.

Reply:
The user need to know this because it is user's job to allocate the memory for the queue obj. If we only exposing the transparent queue handle type `MPIDU_genq_shmem_queue_t`, the user won't know what is size of the queue obj.

Reply:
I see. So why don't we have `MPIDU_genq_shmem_queue_alloc` (which probably can combine with `MPIDU_genq_shmem_queue_init`) ?

Reply:
The user may want to allocated queues on shared memory collective (e.g. pt2pt recv queue use case) or they may want to allocated the queues independently (e.g. SHM collective use case).

If queue is tied to the shmem pool, it would be forcing the collective allocation model.

Reply:
I didn't mean it to be tied to the shmem pool. The shmem queue has to be allocated collectively on shared memory anyway, so instead of user call `MPIDU_Init_shm_alloc`, which not `MPIDU_genq_shmem_queue_init` call it for the user in stead?

Reply:
> If queue is tied to the shmem pool, it would be forcing the collective allocation model.

Oh, just realized that `pool` is already tied to the queue. It is a parameter to `enqueue` `dequeue` since the base address on different processors are different.

Reply:
The queue object need to be on shared memory. The shared memory need to be collectively allocated. If we do a queue allocation function, that function (assuming it allocates shmem) to be collective.

By only initializing a queue obj on an allocated shared memory region is more flexible in that we can init a new queue object without other process participating.

Reply:
The collective allocation of queue might makes sense for recv queues. But it is to strict for SHM collectives queues.

Reply:
> 
> 
> The collective allocation of queue might makes sense for recv queues. But it is too strict for SHM collectives queues.

Pre-mature optimization is the root of all evil. How about let's not worry about collective shmem until that is being worked at? The worst case, we always can create new interface, e.g `MPIDU_genq_shemem_queue_get_size()`, or we can rework current API.

Reply:
We shouldn't do that because we will be removing it. Think about this: if process 0 is going to allocated a shmem queue while process 1 is not going to create a shmem queue, why should we force process 1 to participate the queue allocation? We could simply let it init from a preallocated slab.

Reply:
For example, SHM bcast would be the root enqueue the data into to its own queue, notify everyone to read the cell enqueued. Only the root process needs to init the queue. The other process should not participating the creation of that queue.

Reply:
> why should we force process 1 to participate the queue allocation?

That is a straw man argument. When we have that situation, we can work at the solution then. Let's not to base our design on something that may or may not be important.

Reply:
> 
> 
> For example, SHM bcast would be the root enqueue the data into to its own queue, notify everyone to read the cell enqueued. Only the root process needs to init the queue. The other process should not participating the creation of that queue.

We can't init the queue at init time?

Reply:
Unfortunately no, there is going to be per-comm.

Reply:
If you want to talk about collective shmem, we can discuss offline. I don't think collective shmem as you hinted has any merit, and definitely we shouldn't use some very vague future code to make current argument.